File: skills/llama-parse/scripts/parse.py
#!/usr/bin/env python3
"""
LlamaParse Helper Script (Modernized)
Parses PDF, Word, PowerPoint, Excel, and image documents into Markdown or JSON
using the new unified llama-cloud SDK.
"""

import os
import sys
import time
import argparse
import json
import urllib.request
import urllib.parse
import warnings

# Suppress warnings to keep console output clean
warnings.filterwarnings("ignore")

def check_dependencies():
    missing = []
    try:
        import llama_cloud
    except ImportError:
        missing.append("llama-cloud")
    
    if missing:
        print(f"Error: Missing dependencies: {', '.join(missing)}", file=sys.stderr)
        print("Please install them using: pip install llama-cloud", file=sys.stderr)
        sys.exit(1)

def print_free_plan_usage(api_key, prefix=""):
    try:
        # 1. Get organization ID
        orgs_req = urllib.request.Request(
            "https://api.cloud.llamaindex.ai/api/v1/organizations",
            headers={"Authorization": f"Bearer {api_key}"}
        )
        with urllib.request.urlopen(orgs_req) as response:
            orgs = json.loads(response.read().decode())
            if not orgs:
                return
            org_id = orgs[0]["id"]
            
        # 2. Get usage details
        usage_req = urllib.request.Request(
            f"https://api.cloud.llamaindex.ai/api/v1/organizations/{org_id}/usage",
            headers={"Authorization": f"Bearer {api_key}"}
        )
        with urllib.request.urlopen(usage_req) as response:
            usage_data = json.loads(response.read().decode())
            
        # 3. Extract and display
        free_credits = usage_data.get("usage", {}).get("active_free_credits_usage", [])
        if free_credits:
            start = free_credits[0].get("starting_balance", 0)
            remain = free_credits[0].get("remaining_balance", 0)
            used = start - remain
            print(f"{prefix}[LlamaParse Free Plan Usage]: Monthly limit is {start} credits. Currently used: {used} credits ({remain} remaining).", file=sys.stderr)
            
            # Warn if remaining credits are below 100
            if remain < 100:
                print(f"[WARNING] LlamaParse remaining credits are below 100! Current remaining: {remain} credits.", file=sys.stderr)
    except Exception:
        # Gracefully ignore if we cannot fetch billing info
        pass

def main():
    parser = argparse.ArgumentParser(
        description="Parse documents to Markdown or JSON using the unified LlamaCloud API."
    )
    parser.add_argument(
        "input_files", 
        nargs="*", 
        help="One or more paths to the documents to parse (PDF, DOCX, PPTX, XLSX, PNG, etc.)"
    )
    parser.add_argument(
        "-o", "--output", 
        help="Path to output file or directory. If a directory is specified, outputs are written there with matching basenames."
    )
    parser.add_argument(
        "-k", "--api-key", 
        help="LlamaCloud API key. If omitted, reads LLAMA_CLOUD_API_KEY environment variable."
    )
    parser.add_argument(
        "-t", "--type", 
        choices=["markdown", "text"], 
        default="markdown", 
        help="Result type (default: markdown)"
    )
    parser.add_argument(
        "--tier",
        choices=["fast", "cost_effective", "agentic", "agentic_plus"],
        default="agentic",
        help="Parsing tier (default: agentic). Replaces older --use-vendor-model flag."
    )
    parser.add_argument(
        "-i", "--instruction", 
        help="Custom parsing instruction (prompt) to guide the parser"
    )
    parser.add_argument(
        "-l", "--language", 
        help="Language code for OCR (e.g., 'en', 'de', 'ch_sim' for simplified Chinese)"
    )
    parser.add_argument(
        "-p", "--pages", 
        help="Target pages to parse (comma-separated page numbers/ranges, e.g., '0,2-5')."
    )
    parser.add_argument(
        "-j", "--json", 
        action="store_true", 
        help="Output raw JSON results instead of text/markdown"
    )
    parser.add_argument(
        "-d", "--download-images",
        action="store_true",
        help="Download extracted images/screenshots to the output directory"
    )
    parser.add_argument(
        "--use-vendor-model", 
        action="store_true", 
        help="Deprecated flag (mapped to tier='agentic')"
    )
    parser.add_argument(
        "--vendor-model-name", 
        help="Deprecated flag. Model selection is now handled on the server side via the tier configuration."
    )
    parser.add_argument(
        "--show-usage",
        action="store_true",
        help="Only query and print the LlamaParse Free Plan credit usage, then exit."
    )
    parser.add_argument(
        "-v", "--verbose", 
        action="store_true", 
        help="Print verbose logs"
    )
    
    args = parser.parse_args()
    
    # 1. Check dependencies
    check_dependencies()
    
    # 2. Check API Key
    api_key = args.api_key or os.environ.get("LLAMA_CLOUD_API_KEY")
    if not api_key:
        print("Error: LLAMA_CLOUD_API_KEY is not set and no --api-key flag was provided.", file=sys.stderr)
        print("Please obtain an API key from https://cloud.llamaindex.ai/ and configure it.", file=sys.stderr)
        sys.exit(1)
        
    # If only showing usage
    if args.show_usage:
        print_free_plan_usage(api_key)
        sys.exit(0)
        
    # Check if input files exist when not showing usage
    if not args.input_files:
        parser.error("the following arguments are required: input_files (or use --show-usage)")
        
    for f in args.input_files:
        if not os.path.exists(f):
            print(f"Error: Input file '{f}' does not exist.", file=sys.stderr)
            sys.exit(1)
            
    # Handle deprecated flags
    tier = args.tier
    if args.use_vendor_model:
        if args.verbose:
            print("[Info] Deprecated flag --use-vendor-model mapped to tier='agentic'.", file=sys.stderr)
        tier = "agentic"
    if args.vendor_model_name and args.verbose:
        print(f"[Warning] Deprecated flag --vendor-model-name '{args.vendor_model_name}' ignored. Model selection is handled via --tier.", file=sys.stderr)

    from llama_cloud.client import LlamaCloud
    from llama_cloud import (
        LlamaParseAgenticOptions,
        LlamaParseInputOptions,
        LlamaParseOutputOptions,
        LlamaParseProcessingOptions,
        LlamaParseOcrParameters,
        LlamaParsePageRanges
    )
    
    # 3. Initialize LlamaCloud Client
    client = LlamaCloud(token=api_key)
    
    try:
        results = []
        for input_file in args.input_files:
            if args.verbose:
                print(f"Uploading '{input_file}' to Llama Cloud...", file=sys.stderr)
                
            # Step 1: Upload file to get file_id
            with open(input_file, "rb") as f:
                file_obj = client.files.upload_file(upload_file=f)
            file_id = file_obj.id
            
            if args.verbose:
                print(f"Uploaded successfully. File ID: {file_id}. Starting parse job...", file=sys.stderr)
                
            # Step 2: Configure parsing options
            agentic_opts = None
            if args.instruction:
                agentic_opts = LlamaParseAgenticOptions(custom_prompt=args.instruction)
                
            processing_opts = None
            if args.language:
                processing_opts = LlamaParseProcessingOptions(
                    ocr_parameters=LlamaParseOcrParameters(languages=[args.language])
                )
                
            page_ranges = None
            if args.pages:
                page_ranges = LlamaParsePageRanges(target_pages=args.pages)
                
            output_opts = None
            if args.download_images:
                output_opts = LlamaParseOutputOptions(images_to_save=["screenshot"])
                
            # Step 3: Trigger the parsing job
            job = client.v_2.parse_file(
                file_id=file_id,
                tier=tier,
                version="latest",
                agentic_options=agentic_opts,
                processing_options=processing_opts,
                page_ranges=page_ranges,
                output_options=output_opts
            )
            
            # Step 4: Poll for completion
            if args.verbose:
                print(f"Job ID: {job.id}. Polling status...", file=sys.stderr)
                
            expand_fields = ["items"] if args.json else [args.type]
            if args.download_images:
                expand_fields.append("images_content_metadata")
                
            while True:
                # Retrieve status
                res = client.v_2.get_parse_job(job.id, expand=expand_fields)
                status = res.job.status
                
                if args.verbose:
                    print(f"Current status: {status}", file=sys.stderr)
                    
                if status == "COMPLETED":
                    break
                elif status in ("FAILED", "CANCELLED"):
                    raise Exception(f"Parsing job failed/cancelled with status '{status}'. Error: {res.job.error_message}")
                    
                time.sleep(2)
                
            # Step 5: Format parsed content
            if args.json:
                # Fetch raw JSON result
                json_res = client.parsing.get_job_json_result(job.id)
                output_content = json.dumps(json_res.dict(), indent=2, ensure_ascii=False)
            else:
                if args.type == "markdown":
                    output_content = "\n\n".join([page.markdown for page in res.markdown.pages if page.markdown])
                else:
                    output_content = "\n\n".join([page.text for page in res.text.pages if page.text])
                    
            results.append((input_file, output_content, res))
            
            # Download images if requested
            if args.download_images and res.images_content_metadata:
                for img in res.images_content_metadata.images:
                    if img.url:
                        img_name = os.path.basename(img.url)
                        
                        # Determine output path
                        if args.output:
                            if os.path.isdir(args.output):
                                img_path = os.path.join(args.output, img_name)
                            else:
                                img_path = os.path.join(os.path.dirname(args.output) or ".", img_name)
                        else:
                            img_path = img_name
                            
                        if args.verbose:
                            print(f"Downloading image: {img_name} -> {img_path}", file=sys.stderr)
                            
                        try:
                            img_req = urllib.request.Request(
                                img.url,
                                headers={"Authorization": f"Bearer {api_key}"}
                            )
                            with urllib.request.urlopen(img_req) as img_resp:
                                with open(img_path, "wb") as img_file:
                                    img_file.write(img_resp.read())
                        except Exception as img_err:
                            print(f"Warning: Failed to download image {img_name}: {img_err}", file=sys.stderr)
                            
            # Show Free plan usage after each file is parsed
            print_free_plan_usage(api_key, prefix="\n")
            
        # 6. Handle Output Writing
        if args.output:
            if os.path.isdir(args.output):
                for input_file, content, _ in results:
                    base = os.path.splitext(os.path.basename(input_file))[0]
                    ext = ".json" if args.json else ".md" if args.type == "markdown" else ".txt"
                    out_path = os.path.join(args.output, base + ext)
                    with open(out_path, "w", encoding="utf-8") as out_file:
                        out_file.write(content)
                    print(f"Successfully saved parsed content to: {out_path}", file=sys.stderr)
            else:
                # If single output file and multiple inputs, concatenate them
                concatenated = ""
                for idx, (input_file, content, _) in enumerate(results):
                    if len(results) > 1:
                        concatenated += f"<!-- START OF FILE: {input_file} -->\n"
                    concatenated += content
                    if len(results) > 1:
                        concatenated += f"\n<!-- END OF FILE: {input_file} -->\n\n"
                
                with open(args.output, "w", encoding="utf-8") as out_file:
                    out_file.write(concatenated)
                print(f"Successfully saved parsed content to: {args.output}", file=sys.stderr)
        else:
            for idx, (input_file, content, _) in enumerate(results):
                if len(results) > 1:
                    print(f"--- File: {input_file} ---", file=sys.stderr)
                print(content)
                if len(results) > 1 and idx < len(results) - 1:
                    print("\n" + "="*40 + "\n", file=sys.stderr)
                    
    except Exception as e:
        print(f"Error occurred during parsing: {e}", file=sys.stderr)
        sys.exit(1)

if __name__ == "__main__":
    main()

File: skills/llama-parse/scripts/parse.ts
#!/usr/bin/env node
/**
 * LlamaParse Helper Script (TypeScript)
 * Parses PDF, Word, PowerPoint, Excel, and image documents into Markdown or JSON
 * using the unified @llamaindex/llama-cloud SDK.
 */

import LlamaCloud from "@llamaindex/llama-cloud";
import { readFile, writeFile, mkdir } from "fs/promises";
import { basename, extname, join, dirname } from "path";

// Helper to print help/usage
function printHelp(): void {
  console.log(`
Usage: npx tsx parse.ts [options] <input_files...>

Options:
  -o, --output <path>       Path to output file or directory. If a directory, outputs are written there with matching basenames.
  -k, --api-key <key>       LlamaCloud API key. If omitted, reads LLAMA_CLOUD_API_KEY environment variable.
  -t, --type <type>         Result type: "markdown" or "text" (default: "markdown").
  --tier <tier>             Parsing tier: "fast", "cost_effective", "agentic", "agentic_plus" (default: "agentic").
  -i, --instruction <text>  Custom parsing instruction (prompt) to guide the parser.
  -l, --language <lang>     Language code for OCR (e.g. "en", "ch_sim").
  -p, --pages <ranges>      Target pages to parse (comma-separated page numbers/ranges, e.g. "0,2-5").
  -j, --json                Output raw JSON results instead of text/markdown.
  -d, --download-images     Download extracted images/screenshots to the output directory.
  -v, --verbose             Print verbose logs.
  --show-usage              Only query and print the LlamaParse Free Plan credit usage, then exit.
  -h, --help                Show this help message.
`);
}

// Helper to query usage
async function printFreePlanUsage(apiKey: string, prefix = ""): Promise<void> {
  try {
    // 1. Get organization ID
    const orgsRes = await fetch("https://api.cloud.llamaindex.ai/api/v1/organizations", {
      headers: { Authorization: `Bearer ${apiKey}` },
    });
    if (!orgsRes.ok) return;
    const orgs = (await orgsRes.json()) as any[];
    if (!orgs || orgs.length === 0) return;
    const orgId = orgs[0].id;

    // 2. Get usage details
    const usageRes = await fetch(`https://api.cloud.llamaindex.ai/api/v1/organizations/${orgId}/usage`, {
      headers: { Authorization: `Bearer ${apiKey}` },
    });
    if (!usageRes.ok) return;
    const usageData = (await usageRes.json()) as any;

    // 3. Extract and display
    const freeCredits = usageData.usage?.active_free_credits_usage;
    if (freeCredits && freeCredits.length > 0) {
      const start = freeCredits[0].starting_balance ?? 0;
      const remain = freeCredits[0].remaining_balance ?? 0;
      const used = start - remain;
      console.warn(`${prefix}[LlamaParse Free Plan Usage]: Monthly limit is ${start} credits. Currently used: ${used} credits (${remain} remaining).`);
      if (remain < 100) {
        console.warn(`[WARNING] LlamaParse remaining credits are below 100! Current remaining: ${remain} credits.`);
      }
    }
  } catch (err) {
    // Gracefully ignore billing query failures
  }
}

async function main() {
  const args = process.argv.slice(2);
  const inputFiles: string[] = [];
  let output: string | undefined;
  let apiKey = process.env.LLAMA_CLOUD_API_KEY;
  let type: "markdown" | "text" = "markdown";
  let tier: "fast" | "cost_effective" | "agentic" | "agentic_plus" = "agentic";
  let instruction: string | undefined;
  let language: string | undefined;
  let pages: string | undefined;
  let json = false;
  let downloadImages = false;
  let verbose = false;
  let showUsage = false;

  // Parse command line arguments
  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    if (arg === "-h" || arg === "--help") {
      printHelp();
      process.exit(0);
    } else if (arg === "--show-usage") {
      showUsage = true;
    } else if (arg === "-o" || arg === "--output") {
      output = args[++i];
    } else if (arg === "-k" || arg === "--api-key") {
      apiKey = args[++i];
    } else if (arg === "-t" || arg === "--type") {
      const val = args[++i];
      if (val === "text" || val === "markdown") {
        type = val;
      } else {
        console.error(`Error: Invalid type value: ${val}`);
        process.exit(1);
      }
    } else if (arg === "--tier") {
      const val = args[++i];
      if (val === "fast" || val === "cost_effective" || val === "agentic" || val === "agentic_plus") {
        tier = val;
      } else {
        console.error(`Error: Invalid tier: ${val}`);
        process.exit(1);
      }
    } else if (arg === "-i" || arg === "--instruction") {
      instruction = args[++i];
    } else if (arg === "-l" || arg === "--language") {
      language = args[++i];
    } else if (arg === "-p" || arg === "--pages") {
      pages = args[++i];
    } else if (arg === "-j" || arg === "--json") {
      json = true;
    } else if (arg === "-d" || arg === "--download-images") {
      downloadImages = true;
    } else if (arg === "-v" || arg === "--verbose") {
      verbose = true;
    } else if (arg.startsWith("-")) {
      // Deprecated or unknown flags
      if (arg === "--use-vendor-model") {
        if (verbose) console.warn("[Info] Deprecated flag --use-vendor-model mapped to tier='agentic'.");
        tier = "agentic";
      } else if (arg === "--vendor-model-name") {
        i++; // skip model name value
        if (verbose) console.warn("[Warning] Deprecated flag --vendor-model-name ignored. Model selection is handled via --tier.");
      } else {
        console.error(`Error: Unknown option ${arg}`);
        printHelp();
        process.exit(1);
      }
    } else {
      inputFiles.push(arg);
    }
  }

  if (!apiKey) {
    console.error("Error: LLAMA_CLOUD_API_KEY is not set and no --api-key flag was provided.");
    console.error("Please obtain an API key from https://cloud.llamaindex.ai/ and configure it.");
    process.exit(1);
  }

  if (showUsage) {
    await printFreePlanUsage(apiKey);
    process.exit(0);
  }

  if (inputFiles.length === 0) {
    console.error("Error: No input files provided.");
    printHelp();
    process.exit(1);
  }

  // Initialize client
  const client = new LlamaCloud({ apiKey });

  try {
    const results: Array<{ inputFile: string; content: string; res: any }> = [];

    for (const inputFile of inputFiles) {
      if (verbose) console.warn(`Uploading '${inputFile}' to Llama Cloud...`);

      // 1. Upload file
      const fileBuffer = await readFile(inputFile);
      const fileName = basename(inputFile);
      const fileBlob = new File([fileBuffer], fileName);
      const fileObj = await client.files.create({
        file: fileBlob,
        purpose: "parse",
      });
      const fileId = fileObj.id;

      if (verbose) console.warn(`Uploaded successfully. File ID: {fileId}. Starting parse job...`);

      // 2. Build options
      const agenticOptions = instruction ? { custom_prompt: instruction } : undefined;
      const processingOptions = language ? { ocr_parameters: { languages: [language] } } : undefined;
      const pageRanges = pages ? { target_pages: pages } : undefined;
      const outputOptions = downloadImages ? { images_to_save: ["screenshot" as const] } : undefined;

      // 3. Start parsing job
      const job = await client.v_2.parseFile({
        file_id: fileId,
        tier,
        version: "latest",
        agentic_options: agenticOptions,
        processing_options: processingOptions,
        page_ranges: pageRanges,
        output_options: outputOptions,
      });

      if (verbose) console.warn(`Job ID: ${job.id}. Polling status...`);

      // 4. Poll for completion
      const expand: string[] = json ? ["items"] : [type + "_full"];
      if (downloadImages) {
        expand.push("images_content_metadata");
      }

      let res: any;
      while (true) {
        res = await client.v_2.getParseJob(job.id, { expand });
        const status = res.job?.status;

        if (verbose) console.warn(`Current status: ${status}`);

        if (status === "COMPLETED") {
          break;
        } else if (status === "FAILED" || status === "CANCELLED") {
          throw new Error(`Parsing job failed/cancelled with status '${status}'. Error: ${res.job?.error_message}`);
        }

        await new Promise((resolve) => setTimeout(resolve, 2000));
      }

      // 5. Retrieve content
      let outputContent = "";
      if (json) {
        outputContent = JSON.stringify(res.items ?? {}, null, 2);
      } else {
        outputContent = type === "markdown" ? (res.markdown_full ?? "") : (res.text_full ?? "");
      }

      results.push({ inputFile, content: outputContent, res });

      // 6. Download images if requested
      if (downloadImages && res.images_content_metadata?.images) {
        for (const img of res.images_content_metadata.images) {
          if (img.presigned_url || img.url) {
            const downloadUrl = img.presigned_url || img.url;
            const imgName = basename(downloadUrl.split("?")[0]); // Get basename without query params

            let imgPath = imgName;
            if (output) {
              // Simple check if output is a directory (has no extension)
              const hasExt = extname(output) !== "";
              if (!hasExt) {
                imgPath = join(output, imgName);
              } else {
                imgPath = join(dirname(output), imgName);
              }
            }

            if (verbose) console.warn(`Downloading image: ${imgName} -> ${imgPath}`);

            try {
              const response = await fetch(downloadUrl, {
                headers: { Authorization: `Bearer ${apiKey}` },
              });
              if (response.ok) {
                const buffer = Buffer.from(await response.arrayBuffer());
                await writeFile(imgPath, buffer);
              } else {
                console.error(`Warning: Failed to download image ${imgName}. Status: ${response.status}`);
              }
            } catch (imgErr) {
              console.error(`Warning: Failed to download image ${imgName}:`, imgErr);
            }
          }
        }
      }

      await printFreePlanUsage(apiKey, "\n");
    }

    // 7. Handle outputs writing
    if (output) {
      const hasExt = extname(output) !== "";
      if (!hasExt) {
        // Output path is a directory
        await mkdir(output, { recursive: true });
        for (const { inputFile, content } of results) {
          const baseName = basename(inputFile, extname(inputFile));
          const ext = json ? ".json" : type === "markdown" ? ".md" : ".txt";
          const outPath = join(output, baseName + ext);
          await writeFile(outPath, content, "utf-8");
          console.error(`Successfully saved parsed content to: ${outPath}`);
        }
      } else {
        // Output path is a single file (concatenate if multiple files parsed)
        let concatenated = "";
        for (const { inputFile, content } of results) {
          if (results.length > 1) {
            concatenated += `<!-- START OF FILE: ${inputFile} -->\n`;
          }
          concatenated += content;
          if (results.length > 1) {
            concatenated += `\n<!-- END OF FILE: ${inputFile} -->\n\n`;
          }
        }
        await writeFile(output, concatenated, "utf-8");
        console.error(`Successfully saved parsed content to: ${output}`);
      }
    } else {
      // Print to stdout
      for (let idx = 0; idx < results.length; idx++) {
        const { inputFile, content } = results[idx];
        if (results.length > 1) {
          console.error(`--- File: ${inputFile} ---`);
        }
        console.log(content);
        if (results.length > 1 && idx < results.length - 1) {
          console.error("\n" + "=".repeat(40) + "\n");
        }
      }
    }
  } catch (err: any) {
    console.error("Error occurred during parsing:", err.message || err);
    process.exit(1);
  }
}

main();
